### PR TITLE
Subtle changes to TypeComarer needed for Linker.

### PR DESCRIPTION
**Pull request by DarkDimius**
@Odersky those are the changes that are result of our discussion last Friday.

Only 9714565 is new here, it allows NamedTypes to static modules to be subtypes
of ThisTypes of static modules. This was already allowed for TermRefs.
The commit comment has a motivation example from stdlib.

@Odersky please review.

### CODE REVIEW

File: src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -1696,7 +1696,20 @@ object Types {
...
   final class TermRefWithFixedSym(prefix: Type, name: TermName, val fixedSym: TermSymbol) extends TermRef(prefix, name) with WithFixedSym
-  final class TypeRefWithFixedSym(prefix: Type, name: TypeName, val fixedSym: TypeSymbol) extends TypeRef(prefix, name) with WithFixedSym
+  class TypeRefWithFixedSym(prefix: Type, name: TypeName, val fixedSym: TypeSymbol) extends TypeRef(prefix, name) with WithFixedSym
+
+  /* used by linker */
+  final class TypeRefWithFixedSymAndInfo(prefix: Type, name: TypeName, fixedSym: TypeSymbol, val underl: Type) extends TypeRefWithFixedSym(prefix, name, fixedSym) {
+    override def derivedSelect(prefix: Type)(implicit ctx: Context): Type =
+      if (prefix eq this.prefix) this
+      else {
+        ???
```
Comment (odersky) [inline src/dotty/tools/dotc/core/Types.scala]: ???

Comment (DarkDimius) [inline src/dotty/tools/dotc/core/Types.scala]: In linker this didn't yet happen: the prefix that this type starts with is very precise type and it should not loose precision by getting less precise prefix.

Comment (DarkDimius) [inline src/dotty/tools/dotc/core/Types.scala]: If this happens, I would need to start looking into changes to derivedSelect from https://github.com/lampepfl/dotty/pull/1001

**Comment by DarkDimius** [comment]
@odersky please have a look and merge if LGT you.

**Comment by odersky** [comment]
LGTM.

**Comment by retronym** [comment]
For the record, I was working in this space in scalac recently: https://github.com/scala/scala/compare/2.12.x...retronym:ticket/9542-combo

**Pull request merged**